### ISSUE TITLE
Add information whether updated microcode is available for CPU

### ISSUE THREAD

**Issue by e-alfred** [speed47/spectre-meltdown-checker#205]
It would be great to have a separate line showing information whether there is an updated microcode package for the CPU available (for Intel this list should be applicable: https://newsroom.intel.com/wp-content/uploads/sites/11/2018/04/microcode-update-guidance.pdf). Aside from that, the line `  * CPU microcode is known to cause stability problems: ` should show "not applicable" instead of "NO" in case there is no microcode available. The line `You should enable IBPB to complete retpoline as a Variant 2 mitigation` should be appended with "if updated microcode is available for your CPU".

Here is an example of a CPU that won't get microcode updates but the script still shows the misleading lines:

```
$ spectre-meltdown-checker 
Spectre and Meltdown mitigation detection tool v0.37

Checking for vulnerabilities on current system
Kernel is Linux 4.16.0-1-686-pae #1 SMP Debian 4.16.5-1 (2018-04-29) i686
CPU is Intel(R) Celeron(R) M processor          900MHz

Hardware check
* Hardware support (CPU microcode) for mitigation techniques
  * Indirect Branch Restricted Speculation (IBRS)
    * SPEC_CTRL MSR is available:  NO 
    * CPU indicates IBRS capability:  YES  (SPEC_CTRL feature bit)
  * Indirect Branch Prediction Barrier (IBPB)
    * PRED_CMD MSR is available:  NO 
    * CPU indicates IBPB capability:  YES  (SPEC_CTRL feature bit)
  * Single Thread Indirect Branch Predictors (STIBP)
    * SPEC_CTRL MSR is available:  NO 
    * CPU indicates STIBP capability:  YES  (Intel STIBP feature bit)
  * Enhanced IBRS (IBRS_ALL)
    * CPU indicates ARCH_CAPABILITIES MSR availability:  YES 
    * ARCH_CAPABILITIES MSR advertises IBRS_ALL capability:  NO 
  * CPU explicitly indicates not being vulnerable to Meltdown (RDCL_NO):  NO 
  * CPU microcode is known to cause stability problems:  NO  (model 13 stepping 8 ucode 0x20 cpuid 0x6d8)
* CPU vulnerability to the three speculative execution attack variants
  * Vulnerable to Variant 1:  YES 
  * Vulnerable to Variant 2:  YES 
  * Vulnerable to Variant 3:  YES 

CVE-2017-5753 [bounds check bypass] aka 'Spectre Variant 1'
* Mitigated according to the /sys interface:  YES  (Mitigation: __user pointer sanitization)
* Kernel has array_index_mask_nospec (x86):  YES  (1 occurrence(s) found of 32 bits array_index_mask_nospec())
* Kernel has the Red Hat/Ubuntu patch:  NO 
* Kernel has mask_nospec64 (arm):  NO 
> STATUS:  NOT VULNERABLE  (Mitigation: __user pointer sanitization)

CVE-2017-5715 [branch target injection] aka 'Spectre Variant 2'
* Mitigated according to the /sys interface:  YES  (Mitigation: Full generic retpoline)
* Mitigation 1
  * Kernel is compiled with IBRS support:  YES 
    * IBRS enabled and active:  UNKNOWN 
  * Kernel is compiled with IBPB support:  YES 
    * IBPB enabled and active:  NO 
* Mitigation 2
  * Kernel has branch predictor hardening (arm):  NO 
  * Kernel compiled with retpoline option:  YES 
    * Kernel compiled with a retpoline-aware compiler:  YES  (kernel reports full retpoline compilation)
> STATUS:  NOT VULNERABLE  (Full retpoline is mitigating the vulnerability)
You should enable IBPB to complete retpoline as a Variant 2 mitigation

CVE-2017-5754 [rogue data cache load] aka 'Meltdown' aka 'Variant 3'
* Mitigated according to the /sys interface:  NO  (Vulnerable)
* Kernel supports Page Table Isolation (PTI):  NO 
  * PTI enabled and active:  NO 
  * Reduced performance impact of PTI:  NO  (PCID/INVPCID not supported, performance impact of PTI will be significant)
* Running as a Xen PV DomU:  NO 
> STATUS:  VULNERABLE  (PTI is needed to mitigate the vulnerability)

> How to fix: If you're using a distro kernel, upgrade your distro to get the latest kernel available. Otherwise, recompile the kernel with the CONFIG_PAGE_TABLE_ISOLATION option (named CONFIG_KAISER for some kernels), or the CONFIG_UNMAP_KERNEL_AT_EL0 option (for ARM64)

```

**Comment by Matthew-Bradley** [comment]
This may be hard to cover accurately (there are a lot of old CPUs that will never receive much needed microcode updates for variants 2, 3a and 4) but it may be possible, since we probably won't see more microcode come out for variant 2 (and maybe even 3a and 4).

When it comes to completing retpoline-based defenses by having the kernel poke IBPB on context switch (to prevent userspace-userspace attacks): I would caution against anything that could be interpreted as "CPUs that will never receive microcode updates are more secure." The messiness of these bugs's variants and patch landscape promotes easy misinterpretation of safer positions.

**Comment by speed47** [comment]
This is why I didn't want to implement such check at first: it'll almost always be out of date by design, and could drive people in making wrong assumptions if they use a not-completely-up-to-date version of the script.
As we just got a new batch of microcodes from Intel again, I ended up implementing this check partially in latest master, from their latest PDF "microcode update guidance". This is not complete yet, as all the old unsupported CPUs are missing, the future ones too (I need to detect that properly, not yet sure how), and other vendors are not implemented either for now.

**Comment by knweiss** [comment]
@speed47 Well, you could show on which version of the Intel Microcode Guidance the analysis is based on.
```
* CPU microcode is the latest known available version:  YES
 (you have version 0xb00002e and latest known version according to Intel Microcode Guidance, August 8 2018, is 0xb00002e)
```

**Comment by sethbergman** [comment]
I've found this tool to be an invaluable and extremely informative resource. :1st_place_medal: 
Instead of downloading the script via `curl` and running `chmod +x`, I've cloned the repo and run `git fetch origin && git pull origin master` to be certain I'm running the most recent version of the script. It seems like overkill to have the script output the most recent git commit when the version reads `v0.39+`, with the `+` meaning there have been commits to the master branch since it's last release, but maybe I'm missing something here. :thinking:

**Comment by speed47** [comment]
@crepererum Thanks, there were a couple others like this, fixed on master.

**Comment by speed47** [comment]
@sethbergman @hardfalcon Yes the "+" after the version number indicates you've checked out the `master` branch between two official releases. I've added some logic a couple days ago: now *if* the script is checked out using git (and git is installed, obviously), `git describe` is used to amend the version. For example the current local version I have says:
`Spectre and Meltdown mitigation detection tool v0.39-10-gb2f64e1-dirty` (this would be v0.39+ if git wasn't installed or the script was just wget/curl'ed).

@knweiss Added the version of the Intel Microcode Guidance as you suggested. I hope they'll continue to release this from time to time so we can update it. I'd like to find the same kind of document for AMD microcodes but I'm not sure it exists (publicly).

**Comment by knweiss** [comment]
@speed47 Yes, the guidance document was updated with each new Intel microcode release. Unfortunately, there is (AFAICS) no canonical download page that could be used to download and parse the latest edition automatically.

Regarding AMD: AFAIK there's unfortunately no such guide. The place to look is https://www.amd.com/en/corporate/security-updates. It references Operating System / System Manufacturer resources and also links all AMD papers.

However, AMD release new microcodes in the [amd-ucode](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/amd-ucode) directory of the [linux-firmware](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/log/) git repo. Search for _"AMD cpu"_ in the commit messages. The latest commit is [this](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/commit/?id=7518922bd5b98b137af7aaf3c836f5a498e91609) (notice the `WHENCE` file). This can be used to track the latest versions but there's no additional information:
```
Driver: microcode_amd - AMD CPU Microcode Update Driver for Linux

File: amd-ucode/microcode_amd.bin
Version: 2013-07-10
File: amd-ucode/microcode_amd_fam15h.bin
Version: 2018-05-24
File: amd-ucode/microcode_amd_fam16h.bin
Version: 2014-10-28
File: amd-ucode/microcode_amd_fam17h.bin
Version: 2018-05-15
```

FWIW: Using [MCExtractor](https://github.com/platomav/MCExtractor) I was able to extract the following table from these latest four files:

```
$ ./dist/MCE -skip  -mass
╔══════════════════════════════════════════╗
║         MC Extractor v1.20.0 r80         ║
╚══════════════════════════════════════════╝

Type the full folder path : /tmp/linux-firmware/amd-ucode

Found 4 file(s)

Press enter to start

microcode_amd_fam15h.bin (1/4)

╔══════════════════════════════════════════════════════════════╗
║                             AMD                              ║
╟───┬──────────┬──────────┬────────────┬───────┬────────┬──────╢
║ # │  CPUID   │ Revision │    Date    │  Size │ Offset │ Last ║
╟───┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 1 │ 00600F12 │ 0600063E │ 2018-02-07 │ 0xA20 │  0x54  │ Yes  ║
╟───┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 2 │ 00600F20 │ 06000852 │ 2018-02-06 │ 0xA20 │ 0xA7C  │ Yes  ║
╟───┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 3 │ 00610F01 │ 06001119 │ 2012-07-13 │ 0xA20 │ 0x14A4 │  No  ║
╚═══╧══════════╧══════════╧════════════╧═══════╧════════╧══════╝

microcode_amd_fam16h.bin (2/4)

╔══════════════════════════════════════════════════════════════╗
║                             AMD                              ║
╟───┬──────────┬──────────┬────────────┬───────┬────────┬──────╢
║ # │  CPUID   │ Revision │    Date    │  Size │ Offset │ Last ║
╟───┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 1 │ 00700F01 │ 0700010F │ 2014-02-19 │ 0xD60 │  0x34  │  No  ║
╚═══╧══════════╧══════════╧════════════╧═══════╧════════╧══════╝

microcode_amd.bin (3/4)

╔═══════════════════════════════════════════════════════════════╗
║                              AMD                              ║
╟────┬──────────┬──────────┬────────────┬───────┬────────┬──────╢
║ #  │  CPUID   │ Revision │    Date    │  Size │ Offset │ Last ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 1  │ 00100F22 │ 01000083 │ 2008-04-30 │ 0x3C0 │ 0x134  │  No  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 2  │ 00100F20 │ 01000084 │ 2008-05-01 │ 0x3C0 │ 0x4FC  │  No  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 3  │ 00100F62 │ 010000C7 │ 2010-03-11 │ 0x3C0 │ 0x8C4  │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 4  │ 00100F43 │ 010000C8 │ 2010-03-11 │ 0x3C0 │ 0xC8C  │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 5  │ 00100F81 │ 010000D9 │ 2011-10-12 │ 0x3C0 │ 0x1054 │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 6  │ 00100F80 │ 010000DA │ 2011-10-24 │ 0x3C0 │ 0x141C │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 7  │ 00100F41 │ 010000DB │ 2011-10-24 │ 0x3C0 │ 0x17E4 │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 8  │ 00100FA0 │ 010000DC │ 2011-10-24 │ 0x3C0 │ 0x1BAC │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 9  │ 00200F31 │ 02000032 │ 2008-03-06 │ 0x200 │ 0x1F74 │  No  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 10 │ 00300F10 │ 03000027 │ 2011-13-09 │ 0x3C0 │ 0x217C │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 11 │ 00500F10 │ 05000029 │ 2013-01-21 │ 0x620 │ 0x2544 │ Yes  ║
╟────┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 12 │ 00500F20 │ 05000119 │ 2013-01-18 │ 0x620 │ 0x2B6C │ Yes  ║
╚════╧══════════╧══════════╧════════════╧═══════╧════════╧══════╝

microcode_amd_fam17h.bin (4/4)

╔══════════════════════════════════════════════════════════════╗
║                             AMD                              ║
╟───┬──────────┬──────────┬────────────┬───────┬────────┬──────╢
║ # │  CPUID   │ Revision │    Date    │  Size │ Offset │ Last ║
╟───┼──────────┼──────────┼────────────┼───────┼────────┼──────╢
║ 1 │ 00800F12 │ 08001227 │ 2018-02-09 │ 0xC80 │  0xA4  │ Yes  ║
╚═══╧══════════╧══════════╧════════════╧═══════╧════════╧══════╝
```

 - [AMD microcode container file format description](https://xenbits.xen.org/docs/unstable/misc/amd-ucode-container.txt)

**Comment by speed47** [comment]
v0.40 released

**Comment by Matthew-Bradley** [comment]
@speed47 An important note regarding AMD microcode:

AMD does NOT always release the latest version of their microcode to the linux firmware git. The coreboot people have been extracting blobs from OEM BIOS images (in order to obtain things like VGA BIOS binaries that are needed for the BIOS to init graphics chips) and have found that AMD has been shipping more updated microcode to OEMs. They've requested the updated OEM microcode, but AMD has been (and afaik still is) unwilling to provide the latest microcode to coreboot without their signing an NDA, which is quite infeasible. That's left them in a sticky situation for the last few months, as they have the more up-to-date microcode, but can't distribute it until AMD decided to eventually push it to the linux git.

**Comment by informer2016** [comment]
@Matthew-Bradley and @speed47 , we have submitted the microcode update patches for AMD 15h and 16h to coreboot, but they can't be accepted just because AMD hasn't released these updated microcodes officially as a part of linux-firmware.git . However we have created these unofficial coreboot patches + their installation script, so the people who'd like to build a secure coreboot for G505S could get & install them in a convenient secure way - http://dangerousprototypes.com/docs/Lenovo_G505S_hacking#AMD_microcode_updates

By the way, if you have any questions about G505S laptop I'll be happy to answer: this is excellent quadcore laptop whose A10-5750M CPU doesn't contain any Intel ME / AMD PSP hardware backdoors, and because it is AMD its not vulnerable to some Intel-specific vulnerabilities like Meltdown which require performance-degrading patches. Also possible to install 16 GB of RAM there, making it a nice powerful laptop which is also among the Libreboot candidates because the remaining blobs aren't signed/locked there and are not suspicious , and at the moment we're creating the free software alternatives. And, best of all, Lenovo G505S is really cheap: could be found for $100-$150 in good condition. There's also a cosy community around it :wink:
